utils/preferences/src/extensions/unicode/macros.rs

```diff
@@ -0,0 +1,129 @@
+#[macro_export]
+macro_rules! keyword_list {
+    ($name:ident {
+        $($module:ident::$key:ident),* $(,)?
+    }) => {
+        $(
+            mod $module;
+            pub use $module::*;
+        )*
+
+        #[non_exhaustive]
+        #[derive(Debug, Clone, Eq, PartialEq)]
+        pub enum $name {
+            $($key($key)),*
+        }
+
+        impl TryFrom<(&unicode::Key, &unicode::Value)> for $name {
+            type Error = Error;
+
+            fn try_from((key, value): (&unicode::Key, &unicode::Value)) -> Result<Self, Self::Error> {
+                match key {
+                    $(
+                       _ if $key::matches_ue_key(key) => Ok(Self::$key(value.try_into()?)),
+                    )*
+                    _ => Err(Error::UnknownKeyword),
+                }
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! enum_keyword {
+    ($name:ident {
+        $($key:expr => $variant:ident $(($v2:ident) {$($subk:expr => $subv:ident),*})?),* $(,)?
+    }) => {
+        #[non_exhaustive]
+        #[derive(Debug, Clone, Eq, PartialEq)]
+        pub enum $name {
+            $($variant
+              $((Option<$v2>))?
+            ),*,
+            Custom(unicode::Value)
+        }
+
+        impl TryFrom<&unicode::Value> for $name {
+            type Error = super::Error;
+
+            fn try_from(s: &unicode::Value) -> Result<Self, Self::Error> {
+                Ok(match s {
+                    $(
+                        _ if *s == unicode::value!($key) => $name::$variant $(($v2::try_from(s).ok()))?,
+                        $(
+                            $(_ if *s == unicode::value!($key, $subk) => $name::$variant(Some($v2::$subv))),*,
```

Comment:
Nit: This parses the unicode::Value each time over the match statement. It is more efficient to use `writeable_cmp_bytes` so you don't need to parse anything.